File: day9/main.py
from argparse import ArgumentParser
from math import prod
from pathlib import Path


def is_lower(data, x, y):

    p = data[x][y]

    if x + 1 < len(data) and p >= data[x+1][y]:
        return False

    if y + 1 < len(data[x]) and p >= data[x][y+1]:
        return False

    if x > 0 and p >= data[x-1][y]:
        return False

    if y > 0 and p >= data[x][y-1]:
        return False

    return True


def calc_risk_level(points):
    return sum(points)+len(points)


def calc_basin_size(x, y, data):

    visited = set()

    def visit(x, y):

        visited.add((x, y))

        if x < 0 or y < 0 or x >= len(data) or y >= len(data[x]) or data[x][y] == 9:
            return 0

        return 1 + sum([visit(*p) for p in [(x-1, y), (x+1, y), (x, y-1), (x, y+1)] if p not in visited])

    return visit(x, y)


if __name__ == "__main__":

    parser = ArgumentParser()
    parser.add_argument("-i", "--infile", required=False, type=Path, default=Path("input"))
    args = parser.parse_args()

    lines = Path(args.infile).read_text().rstrip("\n").split("\n")
    data = [[int(x) for x in list(row)] for row in lines]

    results = []
    results_loc = []
    for i in range(len(data)):
        for j in range(len(data[i])):
            if is_lower(data, i, j):
                results_loc.append((i, j))
                results.append(data[i][j])

    print(calc_risk_level(results))

    basin_sizes = [calc_basin_size(x, y, data) for x, y in results_loc]
    print(prod(sorted(basin_sizes)[-3:]))

File: day9/main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"sort"
	"strings"
)

type pos struct {
	x int
	y int
}

func isLower(data [][]int, x int, y int) bool {

	p := data[x][y]

	if x+1 < len(data) && p >= data[x+1][y] {
		return false
	}

	if y+1 < len(data[x]) && p >= data[x][y+1] {
		return false
	}

	if x > 0 && p >= data[x-1][y] {
		return false
	}

	if y > 0 && p >= data[x][y-1] {
		return false
	}

	return true
}

func walkBasin(p pos, data [][]int, ch chan int) {
	var visit func(p pos, ch chan int)
	var next pos
	visited := make(map[pos]struct{})

	visit = func(p pos, ch chan int) {

		visited[p] = struct{}{}

		if !(p.x < 0 || p.y < 0 || p.x >= len(data) || p.y >= len(data[p.x]) || data[p.x][p.y] == 9) {

			ch <- 1

			for _, next = range []pos{pos{p.x + 1, p.y}, pos{p.x - 1, p.y}, pos{p.x, p.y + 1}, pos{p.x, p.y - 1}} {
				if _, ok := visited[next]; !ok {
					visit(next, ch)
				}
			}

		}

	}
	visit(p, ch)
	close(ch)

}

func main() {
	var file string
	flag.StringVar(&file, "infile", "input", "Input file")
	flag.Parse()

	content, err := os.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}

	lines := strings.Split(strings.Trim(string(content), "\n"), "\n")
	data := make([][]int, len(lines))

	for i, x := range lines {
		data[i] = make([]int, len(x))
		for j, y := range x {
			data[i][j] = int(y) - '0'
		}

	}

	// Part 1
	var lows []int
	var lows_loc []pos
	for i, row := range data {
		for j, x := range row {
			if isLower(data, i, j) {
				lows = append(lows, x)
				lows_loc = append(lows_loc, pos{i, j})
			}
		}
	}

	risk_score := 0
	for _, x := range lows {
		risk_score += x + 1
	}

	fmt.Println(risk_score)

	// Part 2
	var sizes []int
	for _, p := range lows_loc {
		ch := make(chan int)
		go walkBasin(p, data, ch)
		size := 0
		for v := range ch {
			size += v
		}
		sizes = append(sizes, size)
	}

	sort.Ints(sizes)
	result := 1
	for _, size := range sizes[len(sizes)-3 : len(sizes)] {
		result *= size
	}
	fmt.Println(result)
}
